File: backend/app/services/business_hours.py
"""
Horário de funcionamento da pizzaria.

Estrutura esperada em pizzarias.horario_funcionamento (JSONB):
    {
      "seg": {"abre": "18:00", "fecha": "23:30", "fechado": false},
      "ter": {...}, ... "dom": {...}
    }

Compatível com o formato antigo (valor string livre) — nesse caso NÃO
enforça horário (assume aberto), para não quebrar configs existentes.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any

try:
    from zoneinfo import ZoneInfo
    TZ = ZoneInfo("America/Sao_Paulo")
except Exception:  # pragma: no cover
    TZ = None

# Monday=0 … Sunday=6
WEEKDAY_KEYS = ["seg", "ter", "qua", "qui", "sex", "sab", "dom"]

DEFAULT_FORA_HORARIO = (
    "Olá! No momento estamos *fechados*. 😴\n"
    "Assim que abrirmos, retornamos seu contato. Obrigado pela preferência!"
)


def _to_minutes(hhmm: str | None) -> int | None:
    if not hhmm or ":" not in str(hhmm):
        return None
    try:
        h, m = str(hhmm).split(":")[:2]
        return int(h) * 60 + int(m)
    except (ValueError, TypeError):
        return None


def esta_aberto(
    horarios: dict[str, Any] | None,
    now: datetime | None = None,
    *,
    override: bool | None = None,
) -> bool:
    """
    True se a pizzaria está aberta agora. O override manual tem prioridade
    sobre a agenda. Em caso de config ausente/legada, retorna True.
    """
    if override is not None:
        return override
    if not horarios:
        return True
    now = now or (datetime.now(TZ) if TZ else datetime.now())
    cfg = horarios.get(WEEKDAY_KEYS[now.weekday()])
    if not cfg or isinstance(cfg, str):
        return True  # sem config estruturada para o dia → não enforça
    if cfg.get("fechado"):
        return False
    abre = _to_minutes(cfg.get("abre"))
    fecha = _to_minutes(cfg.get("fecha"))
    if abre is None or fecha is None:
        return True
    agora = now.hour * 60 + now.minute
    if fecha <= abre:
        # Vira a meia-noite (ex.: 18:00–02:00)
        return agora >= abre or agora < fecha
    return abre <= agora < fecha


_NOMES_DIA = {"seg": "Seg", "ter": "Ter", "qua": "Qua", "qui": "Qui",
              "sex": "Sex", "sab": "Sáb", "dom": "Dom"}

MARCADOR_HORARIO = "{horario}"


def formatar_horario(horarios: dict[str, Any] | None) -> str | None:
    """Horário cadastrado em texto para o cliente, agrupando dias iguais seguidos.

        • Seg a Sex: 18:00 às 23:00
        • Sáb: fechado

    Devolve None se não houver horário estruturado (formato antigo/vazio) — aí
    não há o que mostrar sem arriscar informação errada.
    """
    if not isinstance(horarios, dict) or not any(
        isinstance(horarios.get(k), dict) for k in WEEKDAY_KEYS
    ):
        return None

    def rotulo(dia: dict[str, Any] | None) -> str:
        if not isinstance(dia, dict) or dia.get("fechado") or not dia.get("abre") or not dia.get("fecha"):
            return "fechado"
        return f"{dia['abre']} às {dia['fecha']}"

    grupos: list[tuple[str, str, str]] = []  # (primeiro_dia, ultimo_dia, rotulo)
    for k in WEEKDAY_KEYS:
        r = rotulo(horarios.get(k))
        if grupos and grupos[-1][2] == r:
            grupos[-1] = (grupos[-1][0], k, r)
        else:
            grupos.append((k, k, r))

    linhas = []
    for ini, fim, r in grupos:
        dias = _NOMES_DIA[ini] if ini == fim else f"{_NOMES_DIA[ini]} a {_NOMES_DIA[fim]}"
        linhas.append(f"• {dias}: {r}")
    return "\n".join(linhas)


def mensagem_fora_horario(pizz: Any) -> str:
    """Texto enviado quando o cliente escreve fora do horário.

    O horário de VERDADE — o que decide se o bot atende — é `horario_funcionamento`.
    A mensagem é texto livre digitado no painel, e os dois se descolam assim que um
    muda e o outro não: em produção a mensagem dizia "quinta a domingo, das 18 às
    23" enquanto o cadastro abria todo dia às 08:00. Por isso o texto pode conter
    `{horario}`, trocado pelo horário cadastrado, e a mensagem padrão já o inclui.
    """
    msgs = getattr(pizz, "mensagens_status", None) or {}
    texto = msgs.get("fora_horario") or DEFAULT_FORA_HORARIO
    horario = formatar_horario(getattr(pizz, "horario_funcionamento", None))

    if MARCADOR_HORARIO in texto:
        return texto.replace(MARCADOR_HORARIO, horario or "").rstrip()
    if texto == DEFAULT_FORA_HORARIO and horario:
        return f"{texto}\n\nNosso horário:\n{horario}"
    return texto

File: src/lib/businessHours.ts
/**
 * Cálculo de funcionamento da pizzaria (aberto/fechado) no fuso horário do Brasil (America/Sao_Paulo).
 * Compatível com o backend em Python (business_hours.py).
 */

export const WEEKDAY_KEYS = ["dom", "seg", "ter", "qua", "qui", "sex", "sab"] as const;

export interface DiaHorarioConfig {
  abre?: string;
  fecha?: string;
  fechado?: boolean;
}

/** Retorna um objeto Date com a hora atual no fuso de São Paulo / Brasília. */
export function getBrazilDate(): Date {
  try {
    const now = new Date();
    const spString = now.toLocaleString("en-US", { timeZone: "America/Sao_Paulo" });
    return new Date(spString);
  } catch {
    return new Date();
  }
}

function toMinutes(hhmm?: string | null): number | null {
  if (!hhmm || typeof hhmm !== "string" || !hhmm.includes(":")) return null;
  const [h, m] = hhmm.split(":").map(Number);
  if (isNaN(h) || isNaN(m)) return null;
  return h * 60 + m;
}

/**
 * Retorna true se a loja está aberta no momento.
 * - Se overrideManual !== null e !== undefined, ele prevalece imediatamente (true = forçar aberto, false = forçar fechado).
 * - Se overrideManual === null ou undefined, calcula pela grade de horários configurada no fuso de SP.
 */
export function calcEstaAberto(
  horarios?: Record<string, any> | null,
  overrideManual?: boolean | null,
  nowDate?: Date
): boolean {
  if (overrideManual !== null && overrideManual !== undefined) {
    return Boolean(overrideManual);
  }
  if (!horarios || Object.keys(horarios).length === 0) {
    return true;
  }
  const now = nowDate || getBrazilDate();
  const dayKey = WEEKDAY_KEYS[now.getDay()];
  const cfg = horarios[dayKey];

  if (!cfg || typeof cfg !== "object") {
    return true;
  }
  if (cfg.fechado) {
    return false;
  }

  const abre = toMinutes(cfg.abre);
  const fecha = toMinutes(cfg.fecha);
  if (abre === null || fecha === null) {
    return true;
  }

  const agora = now.getHours() * 60 + now.getMinutes();
  if (fecha <= abre) {
    // Vira a meia-noite (ex: 18:00 às 02:00)
    return agora >= abre || agora < fecha;
  }
  return agora >= abre && agora < fecha;
}

/**
 * Retorna texto amigável do próximo horário de abertura (ex: "Abre hoje às 18:00" ou "Abre amanhã às 18:00").
 */
export function getTextoProximaAbertura(horarios?: Record<string, any> | null, nowDate?: Date): string | null {
  if (!horarios) return null;
  const now = nowDate || getBrazilDate();
  const hojeIdx = now.getDay();
  const agoraMinutos = now.getHours() * 60 + now.getMinutes();

  // Verifica hoje
  const hojeCfg = horarios[WEEKDAY_KEYS[hojeIdx]];
  if (hojeCfg && typeof hojeCfg === "object" && !hojeCfg.fechado && hojeCfg.abre) {
    const abreHoje = toMinutes(hojeCfg.abre);
    if (abreHoje !== null && agoraMinutos < abreHoje) {
      return `Abre hoje às ${hojeCfg.abre}`;
    }
  }

  // Procura nos próximos dias
  const diasLabel = ["domingo", "segunda", "terça", "quarta", "quinta", "sexta", "sábado"];
  for (let offset = 1; offset <= 7; offset++) {
    const idx = (hojeIdx + offset) % 7;
    const cfg = horarios[WEEKDAY_KEYS[idx]];
    if (cfg && typeof cfg === "object" && !cfg.fechado && cfg.abre) {
      if (offset === 1) return `Abre amanhã às ${cfg.abre}`;
      return `Abre ${diasLabel[idx]} às ${cfg.abre}`;
    }
  }
  return null;
}
